x/mongo/driver/operation/count.go

```diff
@@ -102,16 +127,63 @@ func (c *Count) Execute(ctx context.Context) error {
 		ServerAPI:         c.serverAPI,
 	}.Execute(ctx, nil)
 
+	// Swallow error if NamespaceNotFound(26) is returned from aggregate on non-existent namespace
+	if err != nil {
+		dErr, ok := err.(driver.Error)
+		if ok && dErr.Code == 26 {
+			err = nil
+		}
+	}
+	return err
 }
 
 func (c *Count) command(dst []byte, desc description.SelectedServer) ([]byte, error) {
-	dst = bsoncore.AppendStringElement(dst, "count", c.collection)
+	// If wire version < 12 (4.9.0), use count command
+	if desc.WireVersion.Max < 12 {
+		dst = bsoncore.AppendStringElement(dst, "count", c.collection)
+		if c.maxTimeMS != nil {
+			dst = bsoncore.AppendInt64Element(dst, "maxTimeMS", *c.maxTimeMS)
+		}
+		if c.query != nil {
+			dst = bsoncore.AppendDocumentElement(dst, "query", c.query)
+		}
+		return dst, nil
+	}
+	// If wire version >= 12 (4.9.0), use aggregate with $collStats
+	header := bsoncore.Value{Type: bsontype.String, Data: bsoncore.AppendString(nil, c.collection)}
+	if c.collection == "" {
+		header = bsoncore.Value{Type: bsontype.Int32, Data: []byte{0x01, 0x00, 0x00, 0x00}}
+	}
+
+	dst = bsoncore.AppendValueElement(dst, "aggregate", header)
+	cursorIdx, cursorDoc := bsoncore.AppendDocumentStart(nil)
 	if c.maxTimeMS != nil {
 		dst = bsoncore.AppendInt64Element(dst, "maxTimeMS", *c.maxTimeMS)
 	}
 	if c.query != nil {
 		dst = bsoncore.AppendDocumentElement(dst, "query", c.query)
 	}
+
+	countPipeline := bson.A{
+		bson.D{
+			{"$collStats", bson.D{
+				{"count", bson.D{}},
+			}},
+		},
+		bson.D{
+			{"$group", bson.D{
+				{"_id", 1},
+				{"n", bson.D{
+					{"$sum", "$count"},
+				}},
+			}},
+		},
+	}
+	_, countPipelineBSON, _ := bson.MarshalValue(countPipeline)
```

Comment:
We generally try to avoid going through a reflection-based marshalling process when serializing commands unless we're appending a user-provided object and don't know what it looks like in raw BSON (e.g. user-provided aggregation pipelines). It's a little more code, but we should be able to get better performance here with something like:

```
collStatsStage := bsoncore.NewDocumentBuilder()...
groupStage := bsoncore.NewDocumentBuilder()...
pipeline := bsoncore.NewArrayBuilder().
    AppendDocument(collStats).
    AppendDocument(group).
    Build()

dst = bsoncore.AppendArrayElement(dst, "pipeline", pipeline)
```

Reply:
Took me a sec but I got the `bsoncore` definition to work; it's about the same amount of code, and I'm sure will be much faster than marshaling every time.